WIP: Add FV method for T8codeMesh

bennibolm opened the pull request:
This PR adds the implementation of a finite volume scheme. The mesh is organized with t8code.

Functionalities:
- 1. and 2. order ((w)ENO) schemes
- MPI parallel using the exchange routines of t8code
- simple (direct neighbors at interfaces) and extended reconstruction stencil (also neighbors "at corners") (the latter is not MPI parallel implemented)
- Paraview visualization directly using a t8code routine
- support of meshes created by hand (quad, hex, triangular, hybrid) and automatically created (only quad and hex)

Visualization is implemented using a t8code routine directly and therefore without .h5 files and Trixi2Vtk.

Short summary about the current status:
- [ ] Life time issue for the GC tracked Julia object used in C. Using mapping `coordinates2mapping` (closure) is cleaned up by Julia's GC, while still used in t8code. -> SegFaults. Current solution: Deactivating the GC during the simulation in elixir. 

TODOs:
- [ ] Use `@threaded` if possible
- [ ] Visualization with Trixi2Vtk?
- [x] Mesh structures on t8code side are not properly finalized automatically. Require finalization at end of elixir (same for DG). BUT: During testing, the mesh is needed for allocation checks. Leads to SegFaults sometimes (see https://github.com/trixi-framework/Trixi.jl/pull/1844#issuecomment-2479254304). Fixed in #2172 

Open things/problems:
- [x] Check if `convergence_test` works parallel. Seems to work in [1c7413d](https://github.com/trixi-framework/Trixi.jl/pull/1844/commits/1c7413d9d4c0ab50cb6f626a3b67baa9b5475abc)
- [ ] Reconstruction stencil on periodic boundaries
- [x] Second order extended reconstruction stencil. Adapt smaller reconstruction stencils
- [ ] MPI parallel with extended reconstruction stencil (Problem: Using data from other ranks. Update: Still a problem after AMR? Now, I use some other rank's data)
- [ ] AMR (https://github.com/bennibolm/Trixi.jl/pull/135)

bennibolm committed:
Add first working first order FV method on T8codeMesh (periodic boundaries)

bennibolm committed:
Add simple routine to create `cmesh`s by hand

bennibolm committed:
fmt

bennibolm committed:
Fix bug in visualization

bennibolm committed:
Relocate function to adapt to "normal" structure

bennibolm committed:
Add tests

bennibolm committed:
Rename Interface Container

bennibolm committed:
Activate tests; clean up

bennibolm committed:
Merge branch 'main' into t8codemesh-fv

bennibolm committed:
Remove comments in elixirs [skip ci]

bennibolm commented on src/meshes/t8code_mesh.jl:
```diff
@@ -1117,3 +1121,456 @@ end
+    end
+end
+
+# Simple meshes
+# Temporary routines to create simple `cmesh`s by hand
+
+# Directly ported from: `src/t8_cmesh/t8_cmesh_examples.c: t8_cmesh_new_periodic_hybrid`.
+function cmesh_new_periodic_hybrid(comm)::t8_cmesh_t
+    n_dims = 2
+    vertices = [ # Just all vertices of all trees. partly duplicated
+        -1.0, -1.0, 0, # tree 0, triangle
+        0, -1.0, 0,
+        0, 0, 0,
+        -1.0, -1.0, 0, # tree 1, triangle
```
The following 400 lines are just by-hand building routines for simple `cmesh`s. This is only temporary and will be deleted in the future.

bennibolm commented on src/meshes/mesh_io.jl:
```diff
@@ -222,9 +222,13 @@ function save_mesh_file(mesh::P4estMesh, output_directory, timestep, mpi_paralle
 
 # TODO: Implement this function as soon as there is support for this in `t8code`.
-function save_mesh_file(mesh::T8codeMesh, output_directory, timestep, mpi_parallel)
-    error("Mesh file output not supported yet for `T8codeMesh`.")
+# function save_mesh_file(mesh::T8codeMesh, output_directory, timestep, mpi_parallel)
+#     error("Mesh file output not supported yet for `T8codeMesh`.")
+
+#     return joinpath(output_directory, "dummy_mesh.h5")
+# end
 
+function save_mesh_file(mesh::T8codeMesh, output_directory, timestep, mpi_parallel)
     return joinpath(output_directory, "dummy_mesh.h5")
 end
```
All those plotting routines are very improvised and are not in the standard Trixi way.
I use the t8code internal routines to directly create `vtu` files inside Trixi.jl.
They probably should be deleted or changed for this PR.

bennibolm commented on src/meshes/t8code_mesh.jl:
```diff
@@ -37,7 +39,9 @@ mutable struct T8codeMesh{NDIMS, RealT <: Real, IsParallel, NDIMSP2, NNODES} <:
         mesh.current_filename = current_filename
+        mesh.unsaved_changes = true # TODO
         mesh.tree_node_coordinates = tree_node_coordinates
+        mesh.max_number_faces = 4 # TODO: How to automatically adapt for other element types, e.g. triangles?
```
Is there a t8code routine that gives the maximum number of faces per element in a forest?

bennibolm replied:
Outdated

bennibolm replied:
Outdated

bennibolm commented on src/solvers/fv_t8code/fv.jl:
```diff
@@ -0,0 +1,250 @@
+        end
+    end
+
+    # TODO: Boundaries
```
I've implemented the non-periodic stuff locally. They are about 150 lines of additional code. Should I add them to this PR?

bennibolm committed:
Change dispatch from T8codeMesh to FV

bennibolm committed:
Remove not-needed lines of code

bennibolm commented on src/callbacks_step/analysis.jl:
```diff
@@ -141,6 +141,49 @@ function AnalysisCallback(mesh, equations::AbstractEquations, solver, cache;
+                          analysis_integrals = union(default_analysis_integrals(equations),
+                                                     extra_analysis_integrals),
+                          RealT = real(solver),
+                          uEltype = eltype(cache.elements[1]), # TODO: this routine is equal to the existing one except this default type here.
```
The existing routine with `solver::dg` uses `uEltype = eltype(cache.elements)` as the default uEltype. Unfortunately, this is not possible here since elements is a vector consisting of containers for every element by itself.
That's why I created an additional routine.
For the future, the `elements` structure in cache is probably not mergeable and should be changed. I added this comment here to keep the issue in mind.

bennibolm replied:
Solved with restructuring in [72573f8](https://github.com/trixi-framework/Trixi.jl/pull/1844/commits/72573f83dc2133156a3c72e8343736027cd20770)

bennibolm commented on src/callbacks_step/analysis_dg2d.jl:
```diff
@@ -149,6 +149,50 @@ function calc_error_norms(func, u, t, analyzer,
     return l2_error, linf_error
 end
 
+function calc_error_norms(func, u, t, analyzer,
+                          mesh, equations,
+                          initial_condition, solver::FV, cache, cache_analysis)
```
This is probably only valid for the mesh type `T8codeMesh`. TODO: Add dispatching

bennibolm commented on src/callbacks_step/analysis_dg2d.jl:
```diff
@@ -231,6 +275,54 @@ function integrate(func::Func, u,
     end
 end
 
+function integrate_via_indices(func::Func, u,
+                               mesh::T8codeMesh, equations,
+                               solver::FV, cache, args...;
+                               normalize = true) where {Func}
```
TODO: Relocate this routine and `calc_error_norms` to `analysis.jl` since it is dimension independent.

bennibolm committed:
Add comment; correct dispatch

bennibolm committed:
Outsource interface flux

bennibolm committed:
Add prolong2boundaries

bennibolm committed:
Add non periodic elixir and flux calculation

bennibolm committed:
fmt

bennibolm committed:
Clean up RealT and uEltype

bennibolm committed:
Fix bug for simulations with mpi by using simple version of `fill_mesh_info!`

bennibolm committed:
remove show routine

bennibolm committed:
Move other boundary calculations to new function

bennibolm committed:
Add MPI tests

bennibolm committed:
Move calculation of interface data to `fill_mesh_info_fv!`

bennibolm committed:
Update comment

bennibolm committed:
Remove elements from init routines for interfaces and boundaries

bennibolm committed:
Remove not-needed code

bennibolm committed:
fmt

bennibolm committed:
Change structure of element container to  structure-of-arrays style

bennibolm committed:
Reduce unnecessary allocations

bennibolm committed:
Rename routines

bennibolm committed:
Merge branch 'main' into t8codemesh-fv

bennibolm committed:
Add RealT to solver

bennibolm committed:
Fix dual faces

bennibolm committed:
Add first working version with linear reconstruction

bennibolm committed:
fmt

bennibolm committed:
Fix bug

bennibolm committed:
Some changes

bennibolm committed:
Add usage of gradient to `prolong2boundaries` [skip ci]

bennibolm committed:
Clean up cmesh routines

bennibolm committed:
Add simple reconstruction stencil

bennibolm committed:
fmt [skip ci]

bennibolm committed:
Simpify code

bennibolm committed:
Calc reconstruction distance in init_stencil; adapt distance for poriodic domains (adapt test)

bennibolm committed:
Adapt mpi test; fmt

bennibolm committed:
Add source terms; Euler source term elixir; tests

bennibolm committed:
fmt

bennibolm committed:
Change order of tests

bennibolm committed:
Merge branch 'main' into t8codemesh-fv

bennibolm committed:
Adapt structure of communication data to allow parallel 2order fv

bennibolm committed:
Remove 2order extended stencil from parallel tests

bennibolm committed:
Merge branch 'main' into t8codemesh-fv

bennibolm commented:
Note: Right now, it is not possible to use the extended reconstruction kernel (for 2order FV) mpi parallel. That's due to the structure of ghost elements in the t8code ghost cell structure. Only direct (=along an edge) neighbors are listed as ghost cells.
Example: Hybrid mesh on 2 mpi ranks. Plotted the cells and ghost cells of rank 0:
![t8code_ghost_cells](https://github.com/trixi-framework/Trixi.jl/assets/74359358/8cc6f2f1-c562-4aed-9e04-ada6b00a2e6b)

bennibolm committed:
Fix bug; Change default in one elixir

bennibolm committed:
Update elixirs

bennibolm committed:
format

bennibolm committed:
Add solution variables to solution_callback

bennibolm committed:
Adapt elixir

bennibolm committed:
Adapt test after 9eb9643

bennibolm committed:
Merge branch 'main' into t8codemesh-fv

bennibolm committed:
Simplify code

bennibolm committed:
Add first version of reconstruction/limiting

bennibolm committed:
fmt

bennibolm commented:
After [c4a4813](https://github.com/trixi-framework/Trixi.jl/pull/1844/commits/c4a4813158f106d56c40f0be7cf989bdde8cbee2), doing some validation:

***Convergence test***
First order with `elixir_advection_basic.jl` (with a periodic quad mesh and time integration method `CarpenterKennedy2N54` and `T=1.0`):
```
convergence_test("examples/t8code_2d_fv/elixir_advection_basic.jl", 7, order=1, initial_refinement_level=2)

l2
scalar
error     EOC
2.33e-01  -
1.50e-01  0.64
8.55e-02  0.81
4.58e-02  0.90
2.37e-02  0.95
1.21e-02  0.98
6.08e-03  0.99

linf
scalar
error     EOC
3.28e-01  -
2.11e-01  0.63
1.21e-01  0.81
6.47e-02  0.90
3.35e-02  0.95
1.71e-02  0.97
8.60e-03  0.99
```

non-periodic domain:
```
convergence_test("../examples/t8code_2d_fv/elixir_advection_nonperiodic.jl", 7, initial_refinement_level=3, order=1)

l2
scalar
error     EOC
1.94e-01  -
1.24e-01  0.64
7.22e-02  0.78
3.92e-02  0.88
2.05e-02  0.93
1.05e-02  0.97
5.32e-03  0.98

linf
scalar
error     EOC
3.30e-01  -
2.11e-01  0.64
1.21e-01  0.81
6.47e-02  0.90
3.35e-02  0.95
1.71e-02  0.97
8.60e-03  0.99
```


***Second order*** with `elixir_advection_basic.jl` (with a periodic quad mesh and time integration method `CarpenterKennedy2N54` and `T=1.0`):
Using parameter `r = 2`, `epsilon = 1.0e-13`
```
convergence_test("examples/t8code_2d_fv/elixir_advection_basic.jl", 6, order=2, initial_refinement_level=1)

l2
scalar
error     EOC
2.54e-01  -
7.64e-02  1.73
1.89e-02  2.02
5.81e-03  1.70
1.74e-03  1.74
5.15e-04  1.75

linf
scalar
error     EOC
3.56e-01  -
1.19e-01  1.58
3.77e-02  1.66
1.42e-02  1.41
5.41e-03  1.39
2.07e-03  1.39
```

When I remove the limiting completely by `lambda = [0.0; 1.0]`, I get
```
l2
scalar
error     EOC
2.09e-01  -
5.31e-02  1.97
9.68e-03  2.46
1.97e-03  2.29
4.59e-04  2.11
1.12e-04  2.03

linf
scalar
error     EOC
2.95e-01  -
7.20e-02  2.03
1.36e-02  2.40
2.78e-03  2.30
6.48e-04  2.10
1.59e-04  2.03
```


***Free stream preservation***:
Simulation of compressible Euler equation with a constant initial condition
```
trixi_include("examples/t8code_2d_fv/elixir_euler_free_stream.jl", tspan=(0.0, 100.0))
```
- `order=1`: `cfl=0.9`, `~7000 time steps`: Errors `<1.0e-15`
- `order=2`: `cfl=0.9`, `~7000 time steps`: Errors `<1.0e-10` (smaller CFL doesn't make a difference) :/
- `order=2`: No Limiting activated, `cfl=0.9`, `~7000 time steps`: Errors `<1.0e-15`

bennibolm committed:
Adapting MPI tests

bennibolm committed:
fmt

bennibolm committed:
Adapt tests

bennibolm commented on src/meshes/t8code_mesh.jl:
```diff
@@ -31,13 +33,17 @@ mutable struct T8codeMesh{NDIMS, RealT <: Real, IsParallel, NDIMSP2, NNODES} <:
         mesh.current_filename = current_filename
+        mesh.unsaved_changes = false # TODO: if set to `true`, the mesh will be saved to a mesh file.
         mesh.tree_node_coordinates = tree_node_coordinates
+        mesh.max_number_faces = 2 * NDIMS # TODO: How to automatically adapt for other element types, e.g. triangles?
```
Is there a t8code routine that gives the maximum number of faces of all elements in a forest?

bennibolm committed:
Add reconstruction support for extended stencil

bennibolm committed:
Adapt cfl number

bennibolm committed:
Merge branch 'main' into t8codemesh-fv

bennibolm committed:
Add finalizing of t8codemesh to fv elixirs

bennibolm committed:
Remove finalizing from fv elixirs due to seg fault during mpi runs

bennibolm commented:
As in the dg elixirs with T8codeMesh, I added `!isinteractive() && finalize(mesh)` to the end of each fv elixir in [0994aab](https://github.com/trixi-framework/Trixi.jl/pull/1844/commits/0994aab1abd88cc35206173e6d29fecdf272e3c9). This caused segmentation fault errors in the mpi tests (see [here](https://github.com/trixi-framework/Trixi.jl/actions/runs/10558294220/job/29256156920#step:7:11472)).
Also, somehow the nonperiodic test case [failed](https://github.com/trixi-framework/Trixi.jl/actions/runs/10558294220/job/29256153847#step:7:2638).
So, I remove it again in [a46520f](https://github.com/trixi-framework/Trixi.jl/pull/1844/commits/a46520f80c036dd679cd3485c1001f9e8fe3da23).

sloede commented:
> As in the dg elixirs with T8codeMesh, I added `!isinteractive() && finalize(mesh)` to the end of each fv elixir in [0994aab](https://github.com/trixi-framework/Trixi.jl/pull/1844/commits/0994aab1abd88cc35206173e6d29fecdf272e3c9). This caused segmentation fault errors in the mpi tests (see [here](https://github.com/trixi-framework/Trixi.jl/actions/runs/10558294220/job/29256156920#step:7:11472)). Also, somehow the nonperiodic test case [failed](https://github.com/trixi-framework/Trixi.jl/actions/runs/10558294220/job/29256153847#step:7:2638). So, I remove it again in [a46520f](https://github.com/trixi-framework/Trixi.jl/pull/1844/commits/a46520f80c036dd679cd3485c1001f9e8fe3da23).

@jmark Could these issues be fixed by your PR we discussed yesterday? Or are they likely to be caused by the same (or similar) causes?

bennibolm commented:
> > As in the dg elixirs with T8codeMesh, I added `!isinteractive() && finalize(mesh)` to the end of each fv elixir in [0994aab](https://github.com/trixi-framework/Trixi.jl/pull/1844/commits/0994aab1abd88cc35206173e6d29fecdf272e3c9). This caused segmentation fault errors in the mpi tests (see [here](https://github.com/trixi-framework/Trixi.jl/actions/runs/10558294220/job/29256156920#step:7:11472)). Also, somehow the nonperiodic test case [failed](https://github.com/trixi-framework/Trixi.jl/actions/runs/10558294220/job/29256153847#step:7:2638). So, I remove it again in [a46520f](https://github.com/trixi-framework/Trixi.jl/pull/1844/commits/a46520f80c036dd679cd3485c1001f9e8fe3da23).
> 
> @jmark Could these issues be fixed by your PR we discussed yesterday? Or are they likely to be caused by the same (or similar) causes?

Actually, the PR we talked about was the reason I added the lines in the first place. After failing, I was not sure, it should work in my examples at all, so I removed it again.
But @jmark and I already talked about this yesterday, and I will add it again and try to figure out what's going on here.

Long story short, I'm aware of the merged PR, and I am in contact with @jmark.

bennibolm committed:
Add tests using triangles and second hybrid mesh

bennibolm committed:
Add comment on domain size; Allow different T8code version

bennibolm committed:
New quad and tri cmesh routines; New FV T8codeMesh interface

bennibolm committed:
Merge branch 'main' into t8codemesh-fv

bennibolm committed:
fmt

bennibolm committed:
Remove `@t8_load_tree_data` from code; pass C_NULL for now

bennibolm committed:
Add comment in nonperiodic tri mesh simulations

bennibolm committed:
Use just `@cfunction` instead of `@t8_analytic_callback`

bennibolm committed:
Update mapping routines

bennibolm committed:
Merge branch 'main' into t8codemesh-fv

bennibolm committed:
fmt

bennibolm committed:
Remove initial_condition; Disable advection_basic tests

bennibolm committed:
Add 3d support for FV using T8codeMesh (#130)

* Add first version of 3d support

* Complete 3d support

* Add 3d elixirs and tests

* Adapt name in elixir

* Reduce initial refinement level of one test

* Simplify NDIMS=3

* Add 3d mpi tests

* Remove initial_condition from rhs; fmt

* Fix parameter in `x_trans_periodic_3d`

* Remove comment

* Some last changes

* fmt

bennibolm committed:
Support hybrid mesh of flexibel domain size

bennibolm committed:
Merge branch 'main' into t8codemesh-fv

bennibolm committed:
Add finalization to elixirs

bennibolm committed:
Merge branch 'main' into t8codemesh-fv

bennibolm committed:
Add T8code to test environment for now

bennibolm commented on test/Project.toml:
```diff
@@ -16,6 +16,7 @@ Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
 Printf = "de0858da-6303-5e67-8744-51eddeeeb8d7"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 StableRNGs = "860ef19b-820b-49d6-a774-d7a799459cd3"
+T8code = "d0cc0030-9a40-4274-8435-baadcfd54fa1"
```
TODO: Remove T8code when possible

bennibolm replied:
Done in [819014f](https://github.com/trixi-framework/Trixi.jl/pull/1844/commits/819014f3a609635d5f86d94ae0f2031fe07b5158)

bennibolm commented:
I again have the problem in the tests, that finalize(mesh) cleans up the mesh but afterward it is still needed to check the allocations.

bennibolm committed:
Disabled gc in elixir; clean up

bennibolm commented:
Short summary about the current status:
1. Life time issue for the GC tracked Julia object used in C. Using mapping `coordinates2mapping` (closure) is cleaned up by Julia's GC, while still used in t8code. -> SegFaults. Current solution: Deactivating the GC during the simulation in elixir. 
2. Mesh structures on t8code side are not properly finalized automatically. Require finalization at end of elixir (same for DG). BUT: During testing, the mesh is needed for allocation checks. Leads to SegFaults sometimes (see https://github.com/trixi-framework/Trixi.jl/pull/1844#issuecomment-2479254304).

bennibolm committed:
Move allocation tests to elixirs

bennibolm commented:
Regarding
> 2. Mesh structures on t8code side are not properly finalized automatically. Require finalization at end of elixir (same for DG). BUT: During testing, the mesh is needed for allocation checks. Leads to SegFaults sometimes (see https://github.com/trixi-framework/Trixi.jl/pull/1844#issuecomment-2479254304).

Moved the allocation tests to the elixir for now.
Most tests run.
Things to monitor: Error in Codecoverage for t8code_fv test (only once/ normal codecov failure?); Failing mpi tests for ios and windows (always?).

bennibolm committed:
Move t8 eclasses to constructor

bennibolm committed:
Remove T8code from test environment

bennibolm commented:
In [d7770ef](https://github.com/trixi-framework/Trixi.jl/pull/1844/commits/d7770ef11c14f7ebc8a6d0dea6f7477210679d8e), there was also an error in the mpi run on linux. It was the often seen error, which looks like this
`AssertionError: length(u_ode) == nvariables(equations) * nelements(mesh, solver, cache)` (https://github.com/trixi-framework/Trixi.jl/actions/runs/12082232552/job/33692896809#step:7:14394)
It is not present in the next run. Nevertheless, somehow this next run stops at the end of tests without coverage (https://github.com/trixi-framework/Trixi.jl/actions/runs/12083207676/job/33695874959?pr=1844#step:7:14742) without any shown reason.

bennibolm committed:
Add using Test to elixir

bennibolm committed:
Merge branch 'main' into t8codemesh-fv

bennibolm committed:
Reactivate allocation tests in test file

bennibolm committed:
Remove alloc tests from elixir

bennibolm committed:
Add free stream test

JoshuaLampert commented:
> I still have the failing mpi tests for macos and windows.

Thanks to @jmark, these are hopefully fixed after merging https://github.com/DLR-AMR/T8code.jl/pull/78 and updating to the latest T8code.jl version.

bennibolm committed:
Update T8code.jl to newest version

bennibolm commented:
> > I still have the failing mpi tests for macos and windows.
> 
> Thanks to @jmark, these are hopefully fixed after merging [DLR-AMR/T8code.jl#78](https://github.com/DLR-AMR/T8code.jl/pull/78) and updating to the latest T8code.jl version.

Unfortunately not :disappointed: Assuming I'm not missing anything

JoshuaLampert commented:
But [this error from before](https://github.com/trixi-framework/Trixi.jl/actions/runs/12162067228/job/33918141181?pr=1844#step:7:11812) disappeared, right? I guess, there were two probably independent issues before.

JoshuaLampert commented:
[These explicit disabling of the GC](https://github.com/trixi-framework/Trixi.jl/pull/1844/files#diff-a5c878f3c332ecbcf3d8a9fae00c40fdcc47c35bfcdaf2a6caa275488356da38R46) should not be necessary anymore after #2172 or are they, @jmark?

bennibolm commented:
> But [this error from before](https://github.com/trixi-framework/Trixi.jl/actions/runs/12162067228/job/33918141181?pr=1844#step:7:11812) disappeared, right? I guess, there were two probably independent issues before.

Oh yes, this can be true.

bennibolm committed:
Merge branch 'main' into t8codemesh-fv

bennibolm committed:
Implement amr (#135)

* Adapt ndofs

* Adapt nelements

* First working version supporting amr (1order and 2order simple reconstruction stencil; only 2d quads)

* Add tests

* fmt

* Fix 3d tests

* Add support for 3d amr

* fmt and comment

* Add mpi infrastructure for mortars

* Add face areas and normals to communication data

* Add 2d and 3d amr tests for mpi parallel runs

* fmt

* Allow rebalancing

* Remove AMR mpi tests

* Reactivate GC in elixir; anable 2d amr test

* Enable 3d amr test

* Reduce allocations for creation of domain data

* Reduce allocation

* Fix allocation in calc_mortar_flux

* Remove `mesh` from `nelements()`

* Move functions around

* Remove commented code

* Reduce allocations

* Adapt last things; Add amr tests with extended reconstruction stencil

bennibolm committed:
Use Base.max instead of max
